core/connection_checker.js

```diff
@@ -0,0 +1,265 @@
+ * compatible.
+ * @param {!Blockly.Connection} one Connection to compare.
+ * @param {!Blockly.Connection} two Connection to compare against.
+ * @return {boolean} True if the connections share a type.
+ * @public
+ */
+Blockly.ConnectionChecker.prototype.doTypeChecks = function(one, two) {
+  var checkArrayOne = one.getCheck();
+  var checkArrayTwo = two.getCheck();
+
+  if (!checkArrayOne || !checkArrayTwo) {
+    // One or both sides are promiscuous enough that anything will fit.
+    return true;
+  }
+  // Find any intersection in the check lists.
+  for (var i = 0; i < checkArrayOne.length; i++) {
+    if (checkArrayTwo.indexOf(checkArrayOne[i]) != -1) {
+      return true;
+    }
+  }
+  // No intersection.
+  return false;
+};
+
+/**
+ * Check whether this connection can be made by dragging.
+ * @param {!Blockly.Connection} one Connection to compare.
+ * @param {!Blockly.Connection} two Connection to compare against.
+ * @return {boolean} True if the connections share a type.
+ * @public
+ */
+Blockly.ConnectionChecker.prototype.doDragChecks = function(one, two) {
+  // Don't consider insertion markers.
+  if (two.getSourceBlock().isInsertionMarker()) {
+    return false;
+  }
+
+  switch (two.type) {
+    case Blockly.PREVIOUS_STATEMENT:
+      return this.canConnectToPrevious_(one, two);
+    case Blockly.OUTPUT_VALUE: {
+      // Don't offer to connect an already connected left (male) value plug to
+      // an available right (female) value plug.
+      if ((two.isConnected() &&
+          !two.targetBlock().isInsertionMarker()) ||
+          one.isConnected()) {
+        return false;
+      }
+      break;
+    }
+    case Blockly.INPUT_VALUE: {
+      // Offering to connect the left (male) of a value block to an already
+      // connected value pair is ok, we'll splice it in.
+      // However, don't offer to splice into an immovable block.
+      if (two.isConnected() &&
+          !two.targetBlock().isMovable() &&
+          !two.targetBlock().isShadow()) {
+        return false;
+      }
+      break;
+    }
+    case Blockly.NEXT_STATEMENT: {
+      // Don't let a block with no next connection bump other blocks out of the
+      // stack.  But covering up a shadow block or stack of shadow blocks is
+      // fine.  Similarly, replacing a terminal statement with another terminal
+      // statement is allowed.
+      if (two.isConnected() &&
+          !one.getSourceBlock().nextConnection &&
+          !two.targetBlock().isShadow() &&
+          two.targetBlock().nextConnection) {
+        return false;
+      }
+      break;
+    }
+    default:
+      // Unexpected connection type.
+      return false;
+  }
+
+  // Don't let blocks try to connect to themselves or ones they nest.
+  if (Blockly.draggingConnections.indexOf(two) != -1) {
+    return false;
+  }
+
+  return true;
+};
+
+/**
+ * Helper function for drag checking.
+ * @param {!Blockly.Connection} one The connection to check, which must be a
```

Comment:
Just a suggestion, instead of `one` / `two`, we could use `a` and `b`. This matches other util comparison methods we have, and is shorter. There's also a couple of places where we're using things like `blockA` / `blockB` already, so that would match there.
Your call.

Reply:
Done.